firestore/collgroupref.go

```diff
@@ -36,3 +46,146 @@ func newCollectionGroupRef(c *Client, dbPath, collectionID string) *CollectionGr
 		},
 	}
 }
+
+// TODO: add a fromProto to compliment toProto for query. With that serialization is
+// very possible. user could also do json.Marshal(toProto())
+
+// GetPartitionedQueries returns a slice of Query objects, each containing a
+// partition of a collection group. partitionCount must be a positive value and
+// the number of returned partitions may be less than the requested number if
+// providing the desired number would result in partitions with very few documents.
+func (cgr CollectionGroupRef) GetPartitionedQueries(ctx context.Context, partitionCount int) ([]Query, error) {
+	qp, err := cgr.getPartitions(ctx, partitionCount)
+	if err != nil {
+		return nil, err
+	}
+	queries := make([]Query, len(qp))
+	for _, part := range qp {
+		queries = append(queries, part.ToQuery())
+	}
+	return queries, nil
+}
+
+// getPartitions returns a slice of QueryPartition objects, describing a start
+// and end range to query a subsection of the collection group. partitionCount
+// must be a positive value and the number of returned partitions may be less
+// than the requested number if providing the desired number would result in
+// partitions with very few documents.
+func (cgr CollectionGroupRef) getPartitions(ctx context.Context, partitionCount int) ([]queryPartition, error) {
+	orderedQuery := cgr.query().OrderBy(DocumentID, Asc)
+
+	if partitionCount <= 0 {
+		return nil, errors.New("a positive partitionCount must be provided")
+	} else if partitionCount == 1 {
+		return []queryPartition{{CollectionGroupQuery: orderedQuery, StartAt: "", EndBefore: ""}}, nil
+	}
+
+	db := cgr.c.path()
+	ctx = withResourceHeader(ctx, db)
+
+	// CollectionGroup Queries need to be ordered by __name__ ASC.
+	query, err := orderedQuery.toProto()
+	if err != nil {
+		return nil, err
+	}
+	structuredQuery := &firestorepb.PartitionQueryRequest_StructuredQuery{
+		StructuredQuery: query,
+	}
+
+	// Uses default PageSize
+	pbr := &firestorepb.PartitionQueryRequest{
+		Parent:         db + "/documents",
+		PartitionCount: int64(partitionCount),
+		QueryType:      structuredQuery,
+	}
+	cursorReferences := make([]*firestorepb.Value, 0, partitionCount)
+	iter := cgr.c.c.PartitionQuery(ctx, pbr)
+	for {
+		cursor, err := iter.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+
+			return nil, fmt.Errorf("GetPartitions: %v", err)
+		}
+		cursorReferences = append(cursorReferences, cursor.Values...)
```

Comment:
Nit: Is `cursor` a proto message? If so, prefer using the accessor `cursor.GetValues()`. If not, ignore this.

Reply:
It is. Thank you!

Reply:
@noahdietz  we have a few others, I won't change them as part of this pr, but if this is generally true I can make alterations
![image](https://user-images.githubusercontent.com/638797/123308147-2a24a080-d4d8-11eb-9621-8247eee088f1.png)

Reply:
Using the accessor protects from accidental referencing of `nil` when a non-primitive field is unset ([code for this accessor](https://github.com/googleapis/go-genproto/blob/b9e0b3197ced/googleapis/firestore/v1/query.pb.go#L525-L530)). In this case,  expanding a `nil` slice is handled safely by the language (like `len` on a `nil` slice), so it's not a huge deal. But with messages (which are represented as struct pointers) and `optional` primitives (which are represented as primitive pointers), it can be a foot gun. 

Unless you are checking for presence or setting a value, its best practice to use the accessor to retrieve the value.